opts: implemet functional parameters the right way

jenstopp opened the pull request:
Signed-off-by: Jens Drenhaus <jens.drenhaus@9elements.com>

Commit:
opts: implemet functional parameters the right way

Signed-off-by: Jens Drenhaus <jens.drenhaus@9elements.com>

jensdrenhaus commented:
WIP: 
Adaption in stboot.go needs to be made to use the reworked API of opts pkg.

RiSKeD commented on opts/certs_test.go:
```diff
@@ -1,59 +1,52 @@
-import (
-	"errors"
-	"testing"
+// func TestDecodePem(t *testing.T) {
+// 	for _, tt := range []struct {
+// 		name    string
+// 		file    string
+// 		wantErr error
+// 	}{
+// 		{
+// 			name:    "SingleCert_not_exist",
+// 			file:    "file-not-exist.pem",
+// 			wantErr: unix.ENOENT,
+// 		},
+// 		{
+// 			name: "SingleCert_success",
+// 			file: "testdata/testcert.pem",
+// 		},
+// 	} {
+// 		t.Run(tt.name, func(t *testing.T) {
+// 			_, err := decodePem(tt.file)
+// 			if err != nil && !errors.Is(err, tt.wantErr) {
+// 				t.Errorf("decodePem(%s)= cert, %q, Want cert, %v", tt.file, err, tt.wantErr)
+// 			}
+// 		})
+// 	}
+// }
```
Why comment out this test and not remove it?

jensdrenhaus replied:
Done

RiSKeD commented on opts/certs_test.go:
```diff
@@ -1,59 +1,52 @@
-		})
-	}
-}
+// func TestCertFileLoader(t *testing.T) {
+// 	for _, tt := range []struct {
+// 		name    string
+// 		file    string
+// 		wantErr error
+// 	}{
+// 		{
+// 			name: "SingleCert_success",
+// 			file: "testdata/testcert.pem",
+// 		},
+// 	} {
+// 		t.Run(tt.name, func(t *testing.T) {
+// 			certLoader := &HTTPSRootsFile{tt.file}
+// 			if _, err := NewOpts(certLoader); err != nil {
+// 				t.Error(err)
+// 			}
+// 		})
+// 		t.Run(tt.name, func(t *testing.T) {
+// 			certLoader := &SigningRootFile{tt.file}
+// 			if _, err := NewOpts(certLoader); err != nil {
+// 				t.Error(err)
+// 			}
+// 		})
+// 	}
+// }
```
Why is this test needed as well?

jensdrenhaus replied:
Gone.

RiSKeD commented on opts/certs.go:
```diff
@@ -2,106 +2,10 @@ package opts
-	return roots, nil
-}
-
 func printLogCerts(certs ...*x509.Certificate) {
```
This function is not used anymore

Commit:
Remove file loader variants.

I/O operations on files can be performed on the caller side.
Providing and API accapting readers is a solid general approch.

Signed-off-by: Jens Drenhaus <jens.drenhaus@9elements.com>

Commit:
adapt main() to use it

Signed-off-by: Jens Drenhaus <jens.drenhaus@9elements.com>

Commit:
remove printLogCert()

Signed-off-by: Jens Drenhaus <jens.drenhaus@9elements.com>

jensdrenhaus commented:
Using `sterror` needs to be done. Maybe separate PR.

RiSKeD reviewed (dismissed):
Looks good to me, nicely refactored.

Foxboron commented on opts/opts_test.go:
```diff
@@ -54,3 +47,321 @@ func TestNewOpts(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	badPEM, err := os.ReadFile("testdata/cert_bad_pem.pem")
```
Isn't it possible to replace all the calls with `os.Open` which returns an `io.Reader`? Do we strictly need the bytes buffer?

Alternative you could do like this 
```go
func readFile(s string) io.Reader {
	f, err := os.ReadFile(s)
	if err != nil {
		t.Fatal(err)
	}
	return bytes.NewBuffer(f)
}

....

		{
			name:       "Successful loading",
			reader:     readFile(goodCertPEM),
			wantLoaded: true,
			errType:    nil,
		},
```
Or something similar. Prevents the duplicate ReadFile segments and I find it a bit cleaner when extending test cases in the future.

jensdrenhaus replied:
Good one :+1:

Foxboron commented:
One small nitpick, rest looks good :)

Commit:
opts: improve opening files in test tables

Signed-off-by: Jens Drenhaus <jens.drenhaus@9elements.com>

Commit:
opts: improve TestNewOpts

Signed-off-by: Jens Drenhaus <jens.drenhaus@9elements.com>

jensdrenhaus commented:
https://git.glasklar.is/system-transparency/core/stboot/-/tree/main/
